From: =?UTF-8?q?Mika=20Pfl=C3=BCger?= <debian@mikapflueger.de>
Date: Sun, 4 Mar 2012 02:17:02 +0100
Subject: postfix policy adjustments: * Allow user_mail_t to transition to
 postfix_master_t for postalias, confined by roles. Uses
 domain_system_change_exemption() for user_mail_t via
 postfix_domtrans_master() which isn't ideal. * Allow
 postfix_local_t to write to mail_spool_t files for locking * Added
 in_unconfined_r() interface and made postfix user domains use it so
 they can be in the role unconfined_r. Ugly but no better solution
 at this time Closes: #592038 #599053 * Allow postfix_local_t to run
 sendmail for programs like vacation

---
 policy/modules/services/mta.if      |    4 ++++
 policy/modules/services/postfix.if  |   14 ++++++++++----
 policy/modules/services/postfix.te  |   27 +++++++++++++++++++++++++--
 policy/modules/system/unconfined.if |   32 ++++++++++++++++++++++++++++++++
 4 files changed, 71 insertions(+), 6 deletions(-)

Index: refpolicy-2.20110726/policy/modules/services/mta.if
===================================================================
--- refpolicy-2.20110726.orig/policy/modules/services/mta.if	2012-06-21 23:36:07.135051072 +1000
+++ refpolicy-2.20110726/policy/modules/services/mta.if	2012-06-21 23:37:09.427719709 +1000
@@ -96,6 +96,8 @@
 
 	miscfiles_read_localization($1_mail_t)
 
+        kernel_read_system_state($1_mail_t)
+
 	optional_policy(`
 		exim_read_log($1_mail_t)
 		exim_append_log($1_mail_t)
@@ -104,6 +106,8 @@
 
 	optional_policy(`
 		postfix_domtrans_user_mail_handler($1_mail_t)
+		# for postalias - role stops unpriv user from doing it
+		postfix_domtrans_master($1_mail_t)
 	')
 
 	optional_policy(`
Index: refpolicy-2.20110726/policy/modules/services/postfix.if
===================================================================
--- refpolicy-2.20110726.orig/policy/modules/services/postfix.if	2012-06-21 23:34:11.689975726 +1000
+++ refpolicy-2.20110726/policy/modules/services/postfix.if	2012-06-21 23:37:09.427719709 +1000
@@ -163,6 +163,9 @@
 	allow postfix_$1_t self:capability dac_override;
 
 	domtrans_pattern(postfix_user_domtrans, postfix_$1_exec_t, postfix_$1_t)
+	optional_policy(`
+		in_unconfined_r(postfix_$1_t)
+	')
 
 	domain_use_interactive_fds(postfix_$1_t)
 ')
@@ -374,6 +377,7 @@
 	')
 
 	domtrans_pattern($1, postfix_master_exec_t, postfix_master_t)
+	domain_system_change_exemption($1)
 ')
 
 ########################################
@@ -454,7 +458,7 @@
 
 #######################################
 ## <summary>
-##	Execute the master postqueue in the caller domain.
+##	Allow the master postqueue to use a fifo of the caller and send sigchld
 ## </summary>
 ## <param name="domain">
 ##	<summary>
@@ -462,12 +466,14 @@
 ##	</summary>
 ## </param>
 #
-interface(`posftix_exec_postqueue',`
+interface(`posftix_run_postqueue',`
 	gen_require(`
-		type postfix_postqueue_exec_t;
+		type postfix_postqueue_t;
 	')
 
-	can_exec($1, postfix_postqueue_exec_t)
+	allow postfix_postqueue_t $1:fd use;
+	allow postfix_postqueue_t $1:fifo_file rw_file_perms;
+	allow postfix_postqueue_t $1:process sigchld;
 ')
 
 ########################################
Index: refpolicy-2.20110726/policy/modules/services/postfix.te
===================================================================
--- refpolicy-2.20110726.orig/policy/modules/services/postfix.te	2012-06-21 23:37:07.415697255 +1000
+++ refpolicy-2.20110726/policy/modules/services/postfix.te	2012-06-21 23:37:09.427719709 +1000
@@ -26,6 +26,9 @@
 postfix_server_domain_template(local)
 mta_mailserver_delivery(postfix_local_t)
 
+# allow postfix_local_t to run programs like vacation that send mail
+mta_sendmail_domtrans(postfix_local_t, postfix_postdrop_t)
+
 # Program for creating database files
 type postfix_map_t;
 type postfix_map_exec_t;
@@ -183,9 +186,11 @@
 ifdef(`distro_redhat',`
 	# for newer main.cf that uses /etc/aliases
 	mta_manage_aliases(postfix_master_t)
-	mta_etc_filetrans_aliases(postfix_master_t)
 ')
 
+mta_etc_filetrans_aliases(postfix_master_t)
+allow postfix_master_t etc_aliases_t:file manage_file_perms;
+
 optional_policy(`
 	cyrus_stream_connect(postfix_master_t)
 ')
@@ -201,6 +206,8 @@
 
 optional_policy(`
 	mysql_stream_connect(postfix_master_t)
+	mysql_stream_connect(postfix_smtpd_t)
+	mysql_stream_connect(postfix_cleanup_t)
 ')
 
 optional_policy(`
@@ -231,6 +238,11 @@
 manage_files_pattern(postfix_bounce_t, postfix_spool_bounce_t, postfix_spool_bounce_t)
 manage_lnk_files_pattern(postfix_bounce_t, postfix_spool_bounce_t, postfix_spool_bounce_t)
 
+# for milters - may be a bug in postfix
+allow postfix_cleanup_t postfix_smtpd_t:fd use;
+allow postfix_cleanup_t postfix_smtpd_t:unix_stream_socket { getattr read write shutdown };
+allow postfix_cleanup_t postfix_smtpd_t:tcp_socket { read write getattr getopt };
+
 ########################################
 #
 # Postfix cleanup local policy
@@ -240,6 +252,7 @@
 
 # connect to master process
 stream_connect_pattern(postfix_cleanup_t, postfix_private_t, postfix_private_t, postfix_master_t)
+write_sock_files_pattern(postfix_virtual_t,postfix_private_t,postfix_private_t)
 
 rw_fifo_files_pattern(postfix_cleanup_t, postfix_public_t, postfix_public_t)
 write_sock_files_pattern(postfix_cleanup_t, postfix_public_t, postfix_public_t)
@@ -273,6 +286,9 @@
 # for .forward - maybe we need a new type for it?
 rw_sock_files_pattern(postfix_local_t, postfix_private_t, postfix_private_t)
 
+# so it can write to the lock file
+mta_rw_spool(postfix_local_t)
+
 allow postfix_local_t postfix_spool_t:file rw_file_perms;
 
 corecmd_exec_shell(postfix_local_t)
@@ -405,6 +421,8 @@
 	dovecot_domtrans_deliver(postfix_pipe_t)
 ')
 
+corecmd_exec_bin(postfix_pipe_t)
+
 optional_policy(`
 	procmail_domtrans(postfix_pipe_t)
 ')
@@ -569,7 +587,7 @@
 ')
 
 optional_policy(`
-	milter_stream_connect_all(postfix_smtp_t)
+	milter_stream_connect_all(postfix_smtpd_t)
 ')
 
 ########################################
@@ -611,6 +629,11 @@
 ')
 
 optional_policy(`
+	dkim_stream_connect(postfix_smtpd_t)
+	dkim_stream_connect(postfix_cleanup_t)
+')
+
+optional_policy(`
 	sasl_connect(postfix_smtpd_t)
 ')
 
Index: refpolicy-2.20110726/policy/modules/system/unconfined.if
===================================================================
--- refpolicy-2.20110726.orig/policy/modules/system/unconfined.if	2012-06-21 23:34:56.266363475 +1000
+++ refpolicy-2.20110726/policy/modules/system/unconfined.if	2012-06-21 23:37:09.427719709 +1000
@@ -270,6 +270,38 @@
 
 ########################################
 ## <summary>
+##	Allow a domain to be in role unconfined_r
+## </summary>
+## <desc>
+##	<p>
+##	Allow the specified domain to be run in the role unconfined_r
+##	This is suitable for domains that are entered indirectly from
+##	unconfined_t
+##	</p>
+##	<p>
+##	Also allow the domain to send sigchld to unconfined_t and use fds
+##	</p>
+## </desc>
+## <param name="domain">
+##	<summary>
+##	Domain to be in unconfined_r
+##	</summary>
+## </param>
+#
+interface(`in_unconfined_r',`
+	gen_require(`
+		type unconfined_t;
+		role unconfined_r;
+	')
+
+	role unconfined_r types $1;
+	allow $1 unconfined_t:process sigchld;
+	allow $1 unconfined_t:fd use;
+	allow $1 unconfined_t:fifo_file { read write getattr };
+')
+
+########################################
+## <summary>
 ##	Inherit file descriptors from the unconfined domain.
 ## </summary>
 ## <param name="domain">
